### DISCUSSION TITLE
[A-L] (23 luglio 2024) Esercizio 2

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#111] [Esami A-L]
**Esercizio 2:** dato un grafo con pesi $p(v)$ sui vertici e pesi $p(e)$ sugli archi, il _peso totale_ di un cammino è la somma dei pesi dei vertici e archi sul cammino. Assumento $p(v) \gt 0$ per ogni vertice $v$ e $p(e) \gt 0$ per ogni arco $e$ , dare lo pseudocodice di un algoritmo $O(nm)$ che prende in input $G,p$ e due vertici $x,y$ e trova il peso minimo di un cammino da $x$ a $y$ .

**Comment by Elia-Belli** [comment]
## Dijkstra
Possiamo modificare Dijkstra aggiornando il peso minimo di un cammino da $x\to y$ considerando anche il peso del vertice in aggiunta al peso dell'arco da cui è raggiunto.
Utilizziamo l'implementazione con l'heap per rientrare nella complessità, migliorandola:
- richiesta: $O(nm) = O(n^3)$
- ottenuta: $O((n+m)\log n) = O(n^2\log n)$ 

```python
def Dijkstra(G,x):
    Dist = [inf]*n
    H = [] #min_heap

    for v in G.V:  #setup heap
	H.insert(v,Dist[v])

    while(H != []):
	v = H.extract_min()

	for u in v.adiacenti:
            if(u in H):
		alt = Dist[v] + p(v,u) + p(u) # modifica
		if(alt < Dist[u]):
		    Dist[u] = alt
		    H.update_key(u,alt)
    return Dist
```

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/b1c80a5f-c7c8-4575-95c0-e715cff4c986)

**Comment by niccolo-fato** [comment]
```py
Esercizio2(G, x, y):
    Dist: array lungo n = |V(G)| = [0, 0, ..., 0]
    R: insieme = {x}
    while (R ≠ V(G)):
        min = ∞
        min_arco = NULL
        for each (u,v) | u ∈ R ∧ v ∈ R:
            if Dist[u] + w(u,v) + w(v) ≤ min:
                min = Dist[u] + w(u,v) + w(v)
                min_arco = (u,v)
        R.add(min_arco[1])
        Dist[min_arco[1]] = min
    return Dist[y]
```
